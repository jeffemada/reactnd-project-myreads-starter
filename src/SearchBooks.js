import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ListBooks from './ListBooks';
import * as BooksAPI from './BooksAPI';

class SearchBooks extends Component {
  state = { books: [] };

  /**
   * Atualiza as prateleiras dos livros que já estão na minha estante.
   * @param {Array} books - lista de livros retornados na pesquisa
   */
  updateBookshelfOfMyBooks = (books) => {
    const { myBooks } = this.props;

    if (myBooks && myBooks.length > 0) {
      myBooks.forEach((myBook) => {
        const book = books.find((book) => book.id === myBook.id);

        // se encontrou livro, atualiza a prateleira
        if (book) {
          book.shelf = myBook.shelf;
        }
      });
    }
  };

  /**
   * Ouvinte do evento de mudança no campo de pesquisa.
   * @param {Event} event - evento
   */
  searchBooks = (event) => {
    const query = event.target.value;

    if (query) {
      BooksAPI.search(query).then((books) => {
        if (Array.isArray(books) && books.length > 0) {
          this.updateBookshelfOfMyBooks(books);
          this.setState(() => ({ books: books }));
        }
      });
    } else if (this.state.books.length > 0) {
      // TODO bug ao remover query rapidamente, não limpa
      this.setState(() => ({ books: [] }));
    }
  };

  /*
    NOTES: The search from BooksAPI is limited to a particular set of search terms.
    You can find these search terms here:
    https://github.com/udacity/reactnd-project-myreads-starter/blob/master/SEARCH_TERMS.md

    However, remember that the BooksAPI.search method DOES search by title or author. So, don't worry if
    you don't find a specific author or title. Every search is limited by search terms.
  */

  render() {
    const { books } = this.state;
    const { onCloseSearch, onRefresh } = this.props;

    return (
      <div className="search-books">
        <div className="search-books-bar">
          <button className="close-search" onClick={onCloseSearch}>
            Close
          </button>
          <div className="search-books-input-wrapper">
            <form>
              <input type="text" placeholder="Search by title or author" onChange={this.searchBooks} />
            </form>
          </div>
        </div>
        <div className="search-books-results">
          <ListBooks books={books} onRefresh={onRefresh} />
        </div>
      </div>
    );
  }
}

SearchBooks.proptypes = {
  onCloseSearch: PropTypes.func.isRequired,
  onRefresh: PropTypes.func.isRequired
};

export default SearchBooks;
