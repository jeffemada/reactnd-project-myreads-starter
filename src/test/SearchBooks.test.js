import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import SearchBooks from '../SearchBooks';

Enzyme.configure({ adapter: new Adapter() });

const myBooks = [
  {
    title: 'Human Factors and Web Development',
    authors: ['Julie Ratner'],
    imageLinks: {
      smallThumbnail:
        'http://books.google.com/books/content?id=sKvi5fVL-uAC&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
      thumbnail:
        'http://books.google.com/books/content?id=sKvi5fVL-uAC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api'
    },
    id: 'sKvi5fVL-uAC',
    shelf: 'read'
  },
  {
    title: 'Pro React',
    authors: ['Cassio de Sousa Antonio'],
    imageLinks: {
      thumbnail:
        'http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api'
    },
    id: 'PKpPCwAAQBAJ',
    shelf: 'wantToRead'
  }
];

it('should render a SearchBooks', () => {
  const wrapper = shallow(
    <SearchBooks myBooks={myBooks} onCloseSearch={jest.fn()} onRefresh={jest.fn()} onLoading={jest.fn()} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('should search books', () => {
  const wrapper = shallow(
    <SearchBooks myBooks={myBooks} onCloseSearch={jest.fn()} onRefresh={jest.fn()} onLoading={jest.fn()} />
  );
  expect(wrapper.state('lastRequest')).toEqual('');
  wrapper.instance().updateQuery({ target: { value: 'web' } });
  expect(wrapper.state('lastRequest')).toBeDefined();
});
