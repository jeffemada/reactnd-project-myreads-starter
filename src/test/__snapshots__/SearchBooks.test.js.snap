// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a SearchBooks 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBooks
    myBooks={
      Array [
        Object {
          "authors": Array [
            "Julie Ratner",
          ],
          "id": "sKvi5fVL-uAC",
          "imageLinks": Object {
            "smallThumbnail": "http://books.google.com/books/content?id=sKvi5fVL-uAC&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api",
            "thumbnail": "http://books.google.com/books/content?id=sKvi5fVL-uAC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
          },
          "shelf": "read",
          "title": "Human Factors and Web Development",
        },
        Object {
          "authors": Array [
            "Cassio de Sousa Antonio",
          ],
          "id": "PKpPCwAAQBAJ",
          "imageLinks": Object {
            "thumbnail": "http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
          },
          "shelf": "wantToRead",
          "title": "Pro React",
        },
      ]
    }
    onCloseSearch={[MockFunction]}
    onLoading={[MockFunction]}
    onRefresh={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="search-books-bar"
        >
          <Link
            className="close-search"
            replace={false}
            to="/"
          >
            Close
          </Link>
          <div
            className="search-books-input-wrapper"
          >
            <input
              onChange={[Function]}
              placeholder="Search by title or author"
              type="text"
            />
          </div>
        </div>,
        <div
          className="search-books-results"
        >
          <BooksGrid
            books={Array []}
            onLoading={[MockFunction]}
            onRefresh={[Function]}
          />
        </div>,
      ],
      "className": "search-books",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="close-search"
              replace={false}
              to="/"
            >
              Close
            </Link>,
            <div
              className="search-books-input-wrapper"
            >
              <input
                onChange={[Function]}
                placeholder="Search by title or author"
                type="text"
              />
            </div>,
          ],
          "className": "search-books-bar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Close",
              "className": "close-search",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Close",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                onChange={[Function]}
                placeholder="Search by title or author"
                type="text"
              />,
              "className": "search-books-input-wrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Search by title or author",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <BooksGrid
            books={Array []}
            onLoading={[MockFunction]}
            onRefresh={[Function]}
          />,
          "className": "search-books-results",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "books": Array [],
            "onLoading": [MockFunction],
            "onRefresh": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="search-books-bar"
          >
            <Link
              className="close-search"
              replace={false}
              to="/"
            >
              Close
            </Link>
            <div
              className="search-books-input-wrapper"
            >
              <input
                onChange={[Function]}
                placeholder="Search by title or author"
                type="text"
              />
            </div>
          </div>,
          <div
            className="search-books-results"
          >
            <BooksGrid
              books={Array []}
              onLoading={[MockFunction]}
              onRefresh={[Function]}
            />
          </div>,
        ],
        "className": "search-books",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="close-search"
                replace={false}
                to="/"
              >
                Close
              </Link>,
              <div
                className="search-books-input-wrapper"
              >
                <input
                  onChange={[Function]}
                  placeholder="Search by title or author"
                  type="text"
                />
              </div>,
            ],
            "className": "search-books-bar",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Close",
                "className": "close-search",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Close",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  onChange={[Function]}
                  placeholder="Search by title or author"
                  type="text"
                />,
                "className": "search-books-input-wrapper",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Search by title or author",
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <BooksGrid
              books={Array []}
              onLoading={[MockFunction]}
              onRefresh={[Function]}
            />,
            "className": "search-books-results",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "books": Array [],
              "onLoading": [MockFunction],
              "onRefresh": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
